### PR TITLE
Dumping Action Server

### PR DESCRIPTION
**Pull request by collinsc**
# Prerequisites
This branch expands on #20 so that branch should be reviewed and merged before this one is tackled.

# Description
I wrote the dumping action server and integrated it with light detection. It takes an empty action message as a goal. 

It required a few things I needed to make:
- __Camera__: I needed a camera to test with so I integrated one of the 1mp 1080p robotics cameras, and calibrated it using the cv_camera node. There is a rear_cam.launch file that reads in my calibration parameters and spins up the camera.
- __On Demand Image Retrieval__: I needed on demand image retrieval from the rear camera stream. I made an image wrapper node that accomplishes this task. It does so by wrapping a topic and storing it as a super efficient, shared pointer, and retrieving that value on request as a ros service. I instantiate it in a rear_cam_wrapper.launch specifically for the purpose of wrapping the rear camera stream. 
- __Fake Aruco Output__: I needed a message to work with to simulate the messages aruco would give me, so I faked one, `Aruco.action` 

The algorithm takes a couple of parameters, see following sections to see their purposse.
- `min linear velocity`: our slow backing speed
- `max linear velocity`: our normal backing speed
- `min angular velocity`: our slow turning speed
- `max angular velocity`: our normal backing speed
- `image service`: the image service

As a precondition it checks to see if we can see the board, if it can't it rejects the goal. (This goes along with a few conversations we had as a team a long time ago, if we need to change it, super easy one line fix).

It starts the light detection mechanism and does the following algorithm until it see's the light:
1. Check for light
  1.1 if detected stop motors & exit
2. Check for preemption
  2.1 If false stop motors & exit
3. Check for Aruco using image service to get most recent data.
  3.1 if we can see aruco back towards the bin at full speed, turning towards the center
  3.2 else back up at minimum speed at current heading
4. rinse. lather. repeat.

# TODO
This relies on aruco, once that is online we need to integrate the aruco action server.

# Testing
This relies on our hardware and cameras, so this again will be pretty hard to test at home. However I can show you how I tested on the jetson. 

__Figure 1: Execution Output__
screenshot from 2018-02-01 18-16-45

On the left terminal in figure 1, you see the output of the action server.
It connects, and accepts a goal, stops and exits with a success status. This exit is triggered by the light detection device.

On the right you see it output of twist msgs on `/cmd_vel`.  As a surrogate for aruco, in the main control loop I am giving it a fake hard coded aruco action server result. You will notice the last packet recieved is and all 0 stop message.

There is no human perceptible latency for this reaction, so mission accomplished.

### CODE REVIEW

File: src/tfr_dumping/src/dumping_action_server.cpp
Comment (RyanBerge) [inline src/tfr_dumping/src/dumping_action_server.cpp:49]: `dump` is being called by Executive, right?  Do we have a plan in place for what to do if we can't see the board here?

It might be good to use a non-empty goal that has some error codes as a return value so that Exec can attempt to take some appropriate action.  Alternatively, I can see an argument that says the Dumping system needs to try and handle this problem somehow.

File: src/tfr_dumping/src/dumping_action_server.cpp
Comment (RyanBerge) [inline src/tfr_dumping/src/dumping_action_server.cpp]: Presumably this is where we'd add the call to the BinController to actually dump?  If so, let's add a `TODO` comment.

**Review by RyanBerge — CHANGES_REQUESTED** [review]
Just had a couple small comments/requests.  I also want to wait until the Aruco PR gets merged to approve this one, for obvious reasons.

File: src/tfr_dumping/src/dumping_action_server.cpp
Comment (collinsc) [reply src/tfr_dumping/src/dumping_action_server.cpp]: Sure will do!

File: src/tfr_dumping/src/dumping_action_server.cpp
Comment (collinsc) [reply src/tfr_dumping/src/dumping_action_server.cpp:49]: You know we haven't talked about it. From what I can think of there are 3 scenarios.

1. We are pointing in the wrong direction and have room to turn: If this is the case navigation did something really wrong, and we need to run localization again. Since executive ran localization the first time, I think executive should do it again.

2. We are pointing the right direction and we are too close to see the marker: then we should not try and localize, because we will hit the wall and just back up blind. 

3. We are pointing the wrong direction and too close: at this point we are just pretty screwed, we can't turn effectively, and don't know where we are, we should have manual control take over.

It's kind of hard to write code to cover all 3 scenarios well, if we end up any of them our positional information might be compromised or untrustworthy, so it would be difficult to determine a clean solution with our current sensor sources. 

What I suggest we can do is make one of the scenarios more likely than the other, depending on where we set the finish line, and optimize for one or the other using the setup we have.

I think it's a little less risky from a collision standpoint to try and run dumping as far away from the obstacle zone, and as close to the bin as possible. The way we have set up dumping and planned it isn't obstacle aware, and might take a really long time to make it obstacle aware.

If we choose to use fiducial odom publisher we could get some pretty good localization data when we are close we have some more tools at our disposal when we can see the bin. We also have another one of those hd  robotics cams, so that would be pretty easy to incorporate that forward facing.

Regardless, Let's talk about it at the meeting and make it next pull request.

Comment (RyanBerge) [reply src/tfr_dumping/src/dumping_action_server.cpp:49]: All three of those scenarios seem pretty unlikely.  If navigation fails enough that we're in this situation, it may be best to switch to tele-op.  I'm not sure.

**Comment by collinsc** [comment]
@RyanBerge I put in that comment. I think any error handling for the dumping action server requires a little team discussion, and might be out of the scope of this request. So let's address it as a team.

File: src/tfr_dumping/src/dumping_action_server.cpp
Comment (collinsc) [reply src/tfr_dumping/src/dumping_action_server.cpp:49]: @RyanBerge Hey we forgot to talk about this at the meeting. 

Here is what I propose. Let's make a maintenence comment here to address this in testing. However I don't think we know enough to anything here yet.

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (AdamuKaapan) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Type to fix later: *It (doesn't have to be done in this PR, this is super not crucial).

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (AdamuKaapan) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Current coding standard is `lowerCamelCase` for all of these

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (AdamuKaapan) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Current coding standard is `lowerCamelCase`

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (AdamuKaapan) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Current coding standard is `lowerCamelCase`

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (AdamuKaapan) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Current coding standard is `lowerCamelCase`

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (AdamuKaapan) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Current coding standard is `lowerCamelCase` for all of these

File: src/tfr_sensor/src/image_topic_wrapper.cpp
Comment (AdamuKaapan) [inline src/tfr_sensor/src/image_topic_wrapper.cpp:46]: Current coding standard is `lowerCamelCase` for these methods

**Review by AdamuKaapan — CHANGES_REQUESTED** [review]
All looks good to me, just found some naming convention issues.

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (RyanBerge) [inline src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Let's go ahead and fold this into `update_control_msg` like we did with the other PR.

File: src/tfr_dumping/src/dumping_action_client.cpp
Comment (RyanBerge) [inline src/tfr_dumping/src/dumping_action_client.cpp]: Why are we sleeping here?

**Review by RyanBerge — CHANGES_REQUESTED** [review]
One small change and one question in addition to @AdamuKaapan's requests.

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (collinsc) [reply src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: Oh sure I got those mixed up, I'll fix that.

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (collinsc) [reply src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: yup yup yup.

File: src/tfr_sensor/src/image_topic_wrapper.cpp
Comment (collinsc) [reply src/tfr_sensor/src/image_topic_wrapper.cpp:46]: yup

File: src/tfr_dumping/include/tfr_dumping/dumping_action_server.h
Comment (collinsc) [reply src/tfr_dumping/include/tfr_dumping/dumping_action_server.h]: sure,

File: src/tfr_dumping/src/dumping_action_client.cpp
Comment (collinsc) [reply src/tfr_dumping/src/dumping_action_client.cpp]: Probably left over from an old manual test case, I'll delete it.

**Comment by collinsc** [comment]
@AdamuKaapan @RyanBerge I incorperated your changes.

File: src/tfr_dumping/src/dumping_action_client.cpp
Comment (collinsc) [reply src/tfr_dumping/src/dumping_action_client.cpp]: I checked it was just leftover test code.

Comment (collinsc) [reply src/tfr_dumping/src/dumping_action_client.cpp]: /removed it

**Review by RyanBerge — APPROVED** [review]

**Review by AdamuKaapan — APPROVED** [review]

**Pull request merged**
